src/Microsoft.NetCore.Analyzers/Core/Tasks/UseValueTasksCorrectly.cs

```diff
@@ -0,0 +1,547 @@
+                            return;
+                        }
+
+                        if (operation.Parent is IArgumentOperation arg && arg.Kind == OperationKind.Argument)
+                        {
+                            // The "0.09% case" is passing the result of a call directly as an argument to another method.
+                            // This could later result in a problem, as now there's a parameter inside the callee that's
+                            // holding on to the awaitable, and it could await it twice... but the caller is still correct,
+                            // and this analyzer does not perform inter-method analysis.
+                            var originalType = arg.Parameter.Type.OriginalDefinition;
+                            if (originalType.Equals(valueTaskType) || originalType.Equals(valueTaskOfTType))
+                            {
+                                // However, it's really only expected when the parameter type is explicitly a ValueTask{<T>};
+                                // if it's just, say, a TValue, we're likely on a bad path, such as storing the instance into
+                                // a collection of some kind, e.g. Dictionary<string, ValueTask>.Add(..., vt).
+                                var originalParameter = arg.Parameter.OriginalDefinition;
+                                if (originalParameter.Type.Kind != SymbolKind.TypeParameter)
+                                {
+                                    return;
+                                }
+                            }
+                        }
+                        else if (operation.Parent is IConditionalOperation)
+                        {
+                            // This is a branch of a ternary, so consider the ternary expression instead.
+                            operation = operation.Parent;
+                            continue;
+                        }
+                        else if (operation.Parent is ISwitchExpressionArmOperation && operation.Parent.Parent is ISwitchExpressionOperation)
+                        {
+                            // This is a case of a switch expression, so consider the switch expression instead.
+                            operation = operation.Parent.Parent;
+                            continue;
+                        }
+                        else if (operation.Parent is IExpressionStatementOperation)
+                        {
+                            // Warn! This is a statement.  The result should have been used.
+                            operationContext.ReportDiagnostic(Diagnostic.Create(UnconsumedDescriptor, invocation.Syntax.GetLocation()));
+                            return;
+                        }
+                        else if (TryGetLocalSymbolAssigned(operation.Parent, out var valueTypeSymbol, out var cfg, out var startingBlock))
```

Comment:
This case could end up being hit multiple times in the same method. We should try and only analyze the cfg once.

Reply:
Yes, that is correct. The best way to implement this would be to create a separate flow analysis, similar to ones in https://github.com/dotnet/roslyn-analyzers/tree/master/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis, and run it once for the entire operation block, which computes the state at individual operations (program points) and/or basic block start for all value task creations within the method body. I will write up a one pager today to help folks start writing such flow analyses.

Reply:
This is hitting a cache in IOperationExtensions.cs.  Doesn't that address the concern?